robot_smach_states/src/robot_smach_states/navigation/guidance.py

```diff
@@ -30,30 +34,183 @@ def _detect_operator_behind_robot(robot, distance=1.0, radius=0.5):
+
+            # Check the distance
+            if abs(entity_relative_pose.p.x()) < self._x_threshold and abs(entity_relative_pose.p.y()) < self._y_threshold:
+                self._passed_furniture_ids.append(entity.id)
+                side = "left" if entity_relative_pose.p.y() >= 0.0 else "right"
+                rospy.logdebug(
+                    "describe passing entity: {}.\t passed entities is now {}".format(room.id, self._passed_room_ids))
+                return "On our {} you can see the {}".format(side, entity.id)
+
+        # no furniture passed, nothing of interest
+        return ""
+
+    def initialize(self):
+        entities = self._robot.ed.get_entities()
+        self._furniture_entities = [entity for entity in entities if entity.is_a("furniture")]
+        self._room_entities = [room for room in entities if room.type == "room"]
+
+        # Match the furniture entities to rooms
+        self._furniture_entities_room = {room: [] for room in self._room_entities}  # map room entities to furniture entities
+        for item in self._furniture_entities:  # type: Entity
+
+            try:
+                room = self.get_room(item._pose.p)
+            except RuntimeError:
+                rospy.logwarn("{} ({}) not in any room".format(item.id, item._pose.p))
+                continue
+            self._furniture_entities_room[room].append(item)
+            rospy.logdebug("{} ({}) is in the {}".format(item.id, item._pose.p, room.id))
+
+        self._passed_room_ids = []  # Will contain the ids of the rooms that are passed
+        self._passed_furniture_ids = []  # Will contain the ids of the furniture that is passed
+
+        # get initial room
+        position = self._robot.base.get_location()
+        try:
+            room = self.get_room(position.frame.p)
+            self._passed_room_ids.append(room.id)
+        except RuntimeError:
+            rospy.logwarn("position ({}) not in any room".format(position.frame.p))
+
+        # get initial entities
+        r = math.sqrt(self._x_threshold*self._x_threshold + self._y_threshold*self._y_threshold)
+        close_entities = self._robot.ed.get_entities(center_point=position.extractVectorStamped(), radius=r)
+        close_furniture_entities = [entity for entity in close_entities if entity.is_a("furniture")]
+        for entity in close_furniture_entities:
+            self._passed_furniture_ids.append(entity.id)
+        rospy.loginfo("reset TourGuide: passed rooms: {}.\t passed entities {}"
+                      .format(self._passed_room_ids, self._passed_furniture_ids))
+
+    def reset(self):
+        """
+        Resets the passed room ids and the passed furniture ids
+        """
+        self._passed_room_ids = []  # Will contain the ids of the rooms that are passed
+        self._passed_furniture_ids = []  # Will contain the ids of the furniture that is passed
+
+    def in_room(self, room, position):
```

Comment:
I think this function can be generalised so that other states could also use it. The question is if that is usefull or not. @jlunenburg

Reply:
Reminder @jlunenburg

Reply:
Fixed differently